### PR TITLE
Filter kernel arguments by intrinsic type in unique_declarations and check data types (for #1047)

### PR DESCRIPTION
**Pull request by TeranIvy**
As said in #1054, this PR introduces checks for a possibility that a `GH_SCALAR` kernel argument has the same `argument_type` but different `data_type` metadata in different kernels within the same Invoke. In #894 this check was introduced for fields. This will fail compilation and it is now caught by these checks for scalar and field arguments.

`unique_declarations` and `unique_declns_by_intent` in `psyGen.py` are modifed to filter kernel arguments by their `intrinsic_type` (optional parameter) in addition to the `argument_type` so that the above check is made easier.

I also used the opportunity to rename the `DynFields` and `DynScalarArgs` to `LFRicFields` and `LFRicScalarArgs`. I added a new test module for scalar arguments and moved a test from `dynamo0p3_test.py`, which is becoming very bloated. I opened issue #1106 to move scalar and field tests from `dynamo0p3_test.py` to their related test modules.

### CODE REVIEW

File: src/psyclone/tests/domain/lfric/lfric_field_test.py [hunk 1]
```diff
@@ -192,18 +192,43 @@ def test_dynfields_call_err():
...
-            "are ['real', 'integer']." in str(err.value))
+        LFRicFields(invoke)._invoke_declarations(ModuleGen(name="my_mod"))
+    assert ("Found unsupported intrinsic types in Invoke "
+            "'invoke_0_testkern_fs_type' declarations for the field "
+            "arguments ['f1']. Supported types are ['real', 'integer']."
+            in str(err.value))
+
+
+def test_field_invoke_uniq_declns_valid_intrinsic():
+    ''' Tests that all valid access valid intrinsic types for user-defined
```
Comment (arporter) [inline src/psyclone/tests/domain/lfric/lfric_field_test.py:207]: s/valid access//

File: src/psyclone/tests/domain/lfric/lfric_scalar_test.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    '''
+    _, invoke_info = parse(
+        os.path.join(BASE_PATH,
+                     "4.16_multikernel_invokes_real_int_scalar_invalid.f90"),
+        api=TEST_API)
+    psy = PSyFactory(TEST_API, distributed_memory=False).create(invoke_info)
+
+    with pytest.raises(GenerationError) as err:
+        _ = psy.gen
+    assert ("At least one scalar (['b']) in Invoke "
```
Comment (arporter) [inline src/psyclone/tests/domain/lfric/lfric_scalar_test.py:99]: s/At least one scalar/Scalar argument(s) ['b'] in Invoke.../

File: src/psyclone/tests/domain/lfric/lfric_scalar_test.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+            "'invoke_real_and_integer_scalars' has different metadata for "
+            "data type (['gh_real', 'gh_integer']) in different kernels. "
+            "This is invalid." in str(err.value))
+
+
+def test_scalar_invoke_uniq_declns_valid_intrinsic():
+    ''' Tests that all valid access valid intrinsic types for
```
Comment (arporter) [inline src/psyclone/tests/domain/lfric/lfric_scalar_test.py:106]: s/valid access//

File: src/psyclone/tests/dynamo0p3_stubgen_test.py [hunk 4]
```diff
@@ -84,27 +84,29 @@ def test_kernel_stub_invalid_iteration_space():
...
-    internal error if it encounters an unrecognised intrinsic type of
-    scalar when generating a kernel stub.
+def test_lfricscalars_stub_err():
+    ''' Check that LFRicScalarArgs._stub_declarations() raises the
+    expected internal error if it encounters an unrecognised data
+    type of a scalar argument when generating a kernel stub.
 
     '''
-    from psyclone.dynamo0p3 import DynScalarArgs
+    from psyclone.dynamo0p3 import LFRicScalarArgs
```
Comment (arporter) [inline src/psyclone/tests/dynamo0p3_stubgen_test.py:93]: Please could you move these imports up to the top.

File: src/psyclone/tests/dynamo0p3_test.py [hunk 5]
```diff
@@ -1985,6 +1977,23 @@ def test_dyninvoke_uniq_declns_inv_type():
             str(excinfo.value))
 
 
+def test_dyninvoke_uniq_declns_intent_inv_intrinsic():
```
Comment (arporter) [inline src/psyclone/tests/dynamo0p3_test.py:1980]: s/inv/invalid/

File: src/psyclone/tests/dynamo0p3_test.py [hunk 6]
```diff
@@ -1985,6 +1977,23 @@ def test_dyninvoke_uniq_declns_inv_type():
...
+    ''' Tests that we raise an error when Invoke.unique_declns_by_intent()
+    is called for an invalid intrinsic type. '''
+    from psyclone.dynamo0p3 import VALID_INTRINSIC_TYPES
```
Comment (arporter) [inline src/psyclone/tests/dynamo0p3_test.py:1983]: Please mv this import to the top.

**Review by arporter — CHANGES_REQUESTED** [review]
Looks good. I was expecting to see less code rather than more though (see inline) but it may be unavoidable. I don't think we should get too hung up on it though as this functionality is due to be replaced by the PSyIR symbol table (maybe within the next ~12 months?).
Coverage is fine.
Tests and examples are OK with compilation.
Reference guide still builds OK.
I've checked for pycodestyle/pylint - see inline.

File: src/psyclone/dynamo0p3.py [hunk 7]
```diff
@@ -3149,97 +3155,156 @@ def initialise(self, parent):
...
+            scal = [arg.declaration_name for arg in self._scalar_args[intent]]
+            rscal = [arg.declaration_name for
+                     arg in self._real_scalars[intent]]
+            iscal = [arg.declaration_name for
+                     arg in self._int_scalars[intent]]
+            # Check for unsupported intrinsic types
+            scal_inv = set(scal) - set(rscal).union(set(iscal))
+            if scal_inv:
+                raise InternalError(
+                    "Found unsupported intrinsic types in Invoke '{0}' "
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/dynamo0p3.py:3218]: Sounds good. Applied.

File: src/psyclone/dynamo0p3.py [hunk 8]
```diff
@@ -3149,97 +3155,156 @@ def initialise(self, parent):
...
-        Insert declarations for all of the scalar arguments.
+        Add declarations for the scalar arguments.
 
-        :param parent: the f2pygen node in which to insert declarations.
+        :param parent: the f2pygen node in which to insert declarations \
+                       (Invoke or Kernel).
         :type parent: :py:class:`psyclone.f2pygen.SubroutineGen`
 
         '''
+        # Create and add declarations
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/dynamo0p3.py:3281]: Yep, removed.

File: src/psyclone/dynamo0p3.py [hunk 9]
```diff
@@ -2682,54 +2687,53 @@ def _invoke_declarations(self, parent):
...
+        # Check for unsupported intrinsic types
+        fld_inv = (set(fld_arg_list) -
+                   set(real_fld_arg_list).union(set(int_fld_arg_list)))
+        if fld_inv:
+            raise InternalError(
+                "Found unsupported intrinsic types in Invoke '{0}' "
+                "declarations for the field arguments {1}. "
+                "Supported types are {2}.".
+                format(self._invoke.name, list(fld_inv),
+                       list(MAPPING_DATA_TYPES.values())))
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/dynamo0p3.py:2721]: Oh yes, I completely forgot to do search and replace after introducing this list. Done.

File: src/psyclone/dynamo0p3.py [hunk 10]
```diff
@@ -3149,97 +3155,156 @@ def initialise(self, parent):
...
+                     arg in self._int_scalars[intent]]
+            # Check for unsupported intrinsic types
+            scal_inv = set(scal) - set(rscal).union(set(iscal))
+            if scal_inv:
+                raise InternalError(
+                    "Found unsupported intrinsic types in Invoke '{0}' "
+                    "declarations for the scalar arguments {1}. "
+                    "Supported types are {2}.".
+                    format(self._invoke.name, list(scal_inv),
+                           list(MAPPING_DATA_TYPES.values())))
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/dynamo0p3.py:3222]: Yep, done throughout and in `dynamo0p3_test.py`.

File: src/psyclone/tests/dynamo0p3_test.py (hunk 5 above)
Comment (TeranIvy) [reply src/psyclone/tests/dynamo0p3_test.py:1980]: Renamed.

File: src/psyclone/tests/dynamo0p3_test.py (hunk 6 above)
Comment (TeranIvy) [reply src/psyclone/tests/dynamo0p3_test.py:1983]: Moved.

File: src/psyclone/tests/domain/lfric/lfric_scalar_test.py (hunk 2 above)
Comment (TeranIvy) [reply src/psyclone/tests/domain/lfric/lfric_scalar_test.py:99]: Changed messages for scalar and field arguments.

File: src/psyclone/tests/domain/lfric/lfric_scalar_test.py (hunk 3 above)
Comment (TeranIvy) [reply src/psyclone/tests/domain/lfric/lfric_scalar_test.py:106]: Thanks for spotting, removed.

File: src/psyclone/tests/dynamo0p3_stubgen_test.py (hunk 4 above)
Comment (TeranIvy) [reply src/psyclone/tests/dynamo0p3_stubgen_test.py:93]: Moved.

File: src/psyclone/tests/domain/lfric/lfric_field_test.py (hunk 1 above)
Comment (TeranIvy) [reply src/psyclone/tests/domain/lfric/lfric_field_test.py:207]: Thanks for spotting, corrected.

File: src/psyclone/dynamo0p3.py [hunk 11]
```diff
@@ -3149,97 +3155,156 @@ def initialise(self, parent):
...
+                    "At least one scalar ({0}) in Invoke '{1}' has different "
+                    "metadata for data type ({2}) in different kernels. "
+                    "This is invalid.".
+                    format(list(scal_multi_type), self._invoke.name,
+                           list(MAPPING_DATA_TYPES.keys())))
+
+        # Create declarations
+        self._create_declarations(parent)
+
+    def _stub_declarations(self, parent):
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/dynamo0p3.py:3238]: I was hoping for that, too, but I couldn't get around it due to how filtering routines are implemented in `psyGen.py`.  We have functions `args_filter` (stand-alone) and `unique_declarations`/`unique_declns_by_intent` (part of the `Invoke` class) that do very similar things. The first function works on instances of `Descriptor` class and is used in the `Kernel` class and classes related to kernel stubs. The second two take instances of `KernelArgument`s. It is basically same functionality but in two different places. As you say in the comment, we are moving towards PSyIR so there is not much sense in refactoring things.

**Review by arporter — APPROVED** [review]
All requested changes have been made. All tests and examples are OK.
All docs build OK.
Issue no. is not mentioned in any TODOs.

**Pull request merged**